// ngx-extended-pdf-viewer/esm2022/lib/options/ngx-console.mjs
export class NgxConsole {
    static init() {
        if (!globalThis['ngxConsoleFilter']) {
            globalThis['ngxConsoleFilter'] = function (_level, _message) {
                return true;
            };
        }
        if (!globalThis['ngxConsole']) {
            globalThis['ngxConsole'] = new NgxConsole();
        }
    }
    log(message, reason) {
        const filter = globalThis['ngxConsoleFilter'] ?? (() => true);
        if (filter('log', message)) {
            if (reason !== undefined) {
                console.log('%s', message, reason);
            }
            else {
                console.log(message);
            }
        }
    }
    error(message, reason) {
        const filter = globalThis['ngxConsoleFilter'] ?? (() => true);
        if (filter('error', message)) {
            if (reason !== undefined) {
                console.error('%s', message, reason);
            }
            else {
                console.error(message);
            }
        }
    }
    warn(message, reason) {
        const filter = globalThis['ngxConsoleFilter'] ?? (() => true);
        if (filter('warn', message)) {
            if (reason !== undefined) {
                console.warn('%s', message, reason);
            }
            else {
                console.warn(message);
            }
        }
    }
}
