// ngx-extended-pdf-viewer/esm2022/lib/toolbar/pdf-rotate-page-cw/pdf-rotate-page-cw.component.mjs
import { Component, Input, effect } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../../pdf-notification-service";
import * as i2 from "@angular/common";
import * as i3 from "../pdf-shy-button/pdf-shy-button.component";
import * as i4 from "../../responsive-visibility";
export class PdfRotatePageCwComponent {
    notificationService;
    changeDetectorRef;
    showRotateCwButton = true;
    disableRotate = true;
    clockwise = true;
    counterClockwise = true;
    PDFViewerApplication;
    constructor(notificationService, changeDetectorRef) {
        this.notificationService = notificationService;
        this.changeDetectorRef = changeDetectorRef;
        effect(() => {
            this.PDFViewerApplication = notificationService.onPDFJSInitSignal();
            if (this.PDFViewerApplication) {
                this.onPdfJsInit();
            }
        });
    }
    rotateCW() {
        this.PDFViewerApplication?.eventBus.dispatch('rotatecw');
    }
    onPdfJsInit() {
        this.PDFViewerApplication?.eventBus.on('updateuistate', (event) => this.updateUIState(event));
    }
    updateUIState(event) {
        this.disableRotate = event.pagesCount === 0;
        this.changeDetectorRef.markForCheck();
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.11", ngImport: i0, type: PdfRotatePageCwComponent, deps: [{ token: i1.PDFNotificationService }, { token: i0.ChangeDetectorRef }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.3.11", type: PdfRotatePageCwComponent, selector: "pdf-rotate-page-cw", inputs: { showRotateCwButton: "showRotateCwButton", clockwise: "clockwise", counterClockwise: "counterClockwise" }, ngImport: i0, template: "<pdf-shy-button\n  *ngIf=\"clockwise\"\n  [action]=\"rotateCW\"\n  primaryToolbarId=\"primaryPageRotateCw\"\n  class=\"rotateCw\"\n  [cssClass]=\"showRotateCwButton | responsiveCSSClass : 'hiddenXLView'\"\n  title=\"Rotate Clockwise\"\n  l10nId=\"pdfjs-page-rotate-cw-button\"\n  l10nLabel=\"pdfjs-page-rotate-cw-button-label\"\n  [disabled]=\"disableRotate\"\n  [order]=\"900\"\n  [closeOnClick]=\"false\"\n  image=\"<svg style='width: 23px; height: 23px' viewBox='0 0 24 24'><path fill='currentColor' d='M12 3C7.03 3 3 7.03 3 12S7.03 21 12 21C14 21 15.92 20.34 17.5 19.14L16.06 17.7C14.87 18.54 13.45 19 12 19C8.13 19 5 15.87 5 12S8.13 5 12 5 19 8.13 19 12H16L20 16L24 12H21C21 7.03 16.97 3 12 3'/></svg>\"\n>\n</pdf-shy-button>\n", styles: [":host:focus{outline:none}button:focus{outline:none}svg:focus{outline:none}button{padding:0;margin-top:0;margin-bottom:0}\n"], dependencies: [{ kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i3.PdfShyButtonComponent, selector: "pdf-shy-button", inputs: ["primaryToolbarId", "secondaryMenuId", "cssClass", "eventBusName", "l10nId", "l10nLabel", "title", "toggled", "disabled", "order", "action", "closeOnClick", "onlySecondaryMenu", "image"] }, { kind: "pipe", type: i4.ResponsiveCSSClassPipe, name: "responsiveCSSClass" }] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.11", ngImport: i0, type: PdfRotatePageCwComponent, decorators: [{
            type: Component,
            args: [{ selector: 'pdf-rotate-page-cw', template: "<pdf-shy-button\n  *ngIf=\"clockwise\"\n  [action]=\"rotateCW\"\n  primaryToolbarId=\"primaryPageRotateCw\"\n  class=\"rotateCw\"\n  [cssClass]=\"showRotateCwButton | responsiveCSSClass : 'hiddenXLView'\"\n  title=\"Rotate Clockwise\"\n  l10nId=\"pdfjs-page-rotate-cw-button\"\n  l10nLabel=\"pdfjs-page-rotate-cw-button-label\"\n  [disabled]=\"disableRotate\"\n  [order]=\"900\"\n  [closeOnClick]=\"false\"\n  image=\"<svg style='width: 23px; height: 23px' viewBox='0 0 24 24'><path fill='currentColor' d='M12 3C7.03 3 3 7.03 3 12S7.03 21 12 21C14 21 15.92 20.34 17.5 19.14L16.06 17.7C14.87 18.54 13.45 19 12 19C8.13 19 5 15.87 5 12S8.13 5 12 5 19 8.13 19 12H16L20 16L24 12H21C21 7.03 16.97 3 12 3'/></svg>\"\n>\n</pdf-shy-button>\n", styles: [":host:focus{outline:none}button:focus{outline:none}svg:focus{outline:none}button{padding:0;margin-top:0;margin-bottom:0}\n"] }]
        }], ctorParameters: () => [{ type: i1.PDFNotificationService }, { type: i0.ChangeDetectorRef }], propDecorators: { showRotateCwButton: [{
                type: Input
            }], clockwise: [{
                type: Input
            }], counterClockwise: [{
                type: Input
            }] } });
